src/yajl/yajl_tree.c

```diff
@@ -0,0 +1,511 @@
+}
+
+static int handle_null (void *ctx)
+{
+    yajl_val v;
+
+    v = value_alloc (yajl_t_null);
+    if (v == NULL) {
+        RETURN_ERROR ((context_t *) ctx, STATUS_ABORT, "Out of memory");
+    }
+
+    return ((context_add_value (ctx, v) == 0) ? STATUS_CONTINUE : STATUS_ABORT);
+}
+
+/*
+ * Public functions
+ */
+yajl_val yajl_tree_parse (const char *input,
+                          char *error_buffer, size_t error_buffer_size)
+{
+    static const yajl_callbacks callbacks = {
+        /* null        = */ handle_null,
+        /* boolean     = */ handle_boolean,
+        /* integer     = */ NULL,
+        /* double      = */ NULL,
+        /* number      = */ handle_number,
+        /* string      = */ handle_string,
+        /* start map   = */ handle_start_map,
+        /* map key     = */ handle_string,
+        /* end map     = */ handle_end_map,
+        /* start array = */ handle_start_array,
+        /* end array   = */ handle_end_array
+    };
+
+    yajl_handle handle;
+    yajl_status status;
+    char *internal_err_str;
+    context_t ctx = { NULL, NULL, NULL, 0 };
+
+    ctx.errbuf = error_buffer;
+    ctx.errbuf_size = error_buffer_size;
+
+    if (error_buffer != NULL) {
+        memset (error_buffer, 0, error_buffer_size);
+    }
+
+    handle = yajl_alloc (&callbacks, NULL, &ctx);
+    yajl_config(handle, yajl_allow_comments, 1);
+
+    status = yajl_parse(handle,
+                        (unsigned char *) input,
+                        strlen (input));
+    status = yajl_complete_parse (handle);
+    if (status != yajl_status_ok) {
+        if (error_buffer != NULL && error_buffer_size > 0) {
+            internal_err_str = (char *) yajl_get_error(handle, 1,
+                               (const unsigned char *) input,
+                               strlen(input));
+            snprintf(error_buffer, error_buffer_size, "%s", internal_err_str);
+            YA_FREE(&(handle->alloc), internal_err_str);
+        }
+        yajl_free (handle);
+        return NULL;
+    }
+
+    yajl_free (handle);
+    return (ctx.root);
+}
+
+yajl_val yajl_tree_get(yajl_val n, const char **path, yajl_type type)
+{
+    if (!path) {
+        return NULL;
+    }
+    while (n && *path) {
+        size_t i;
+        size_t len;
+
+        if (n->type != yajl_t_object) {
+            return NULL;
+        }
+        len = n->u.object.len;
+        for (i = 0; i < len; i++) {
+            if (!strcmp(*path, n->u.object.keys[i])) {
+                n = n->u.object.values[i];
+                break;
+            }
+        }
+        if (i == len) {
+            return NULL;
+        }
+        path++;
+    }
+    if (n && type != yajl_t_any && type != n->type) {
+        n = NULL;
+    }
+    return n;
+}
+
+void yajl_tree_free (yajl_val v)
+{
+    if (v == NULL) {
+        return;
+    }
+
+    if (YAJL_IS_STRING(v)) {
+        memset(v->u.string, 0, strlens(v->u.string));
```

Comment:
I'd still recommend making a fork of that repo in your github org and then
make a submodule of that. So one can check that repo.

Maybe upstream would accept this as a patch even?
On Jul 17, 2015 11:05 PM, "dbitbox" <notifications@github.com> wrote:

> In src/yajl/yajl_tree.c
> <https://github.com/digitalbitbox/mcu/pull/26#discussion_r34931011>:
>
> > +        path++;
> > +    }
> > +    if (n && type != yajl_t_any && type != n->type) {
> > +        n = NULL;
> > +    }
> > +    return n;
> > +}
> > +
> > +void yajl_tree_free (yajl_val v)
> > +{
> > +    if (v == NULL) {
> > +        return;
> > +    }
> > +
> > +    if (YAJL_IS_STRING(v)) {
> > +        memset(v->u.string, 0, strlens(v->u.string));
>
> I added this modification in order to erase json strings before they are
> freed. Therefore, looks like a submodule cannot handle this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/digitalbitbox/mcu/pull/26/files#r34931011>.
>